torch/csrc/jit/tensorexpr/ir.cpp

```diff
@@ -10,46 +10,190 @@ static Dtype ChooseDtype(const Dtype& buffer_dtype, const Dtype& index_dtype) {
+    const ExprHandle& mask) {
+  return ExprHandle(new Load(
+      dtype, buf.node(), ExprHandleVectorToExprVector(indices), mask.node()));
+}
+
 Store::Store(
     const Buffer& buffer,
-    const Expr* index,
+    const std::vector<const Expr*>& indices,
     const Expr* value,
     const Expr* mask)
-    : Store(buffer.data(), index, value, mask) {
+    : Store(buffer.data(), indices, value, mask) {
   if (buffer.dtype().scalar_type() != value->dtype().scalar_type()) {
     throw malformed_input();
   }
 }
 
+Store::Store(
+    const Buf* buf,
+    const std::vector<const Expr*>& indices,
+    const Expr* value,
+    const Expr* mask)
+    : buf_(buf), indices_(indices), value_(value), mask_(mask) {
+  if (buf->dtype() != kHandle) {
+    throw malformed_input();
+  }
+  /*
+  TODO: Reenable the checks.
+  The reason they are disabled is that kernel.cpp is using Buffers somewhat
+  loosely: we don't set dimensions properly and just construct index expressions
+  directly. We should harden that part and then we'd be able to turn on these
+  checks.
+
+  if (!indicesValid(indices)) {
+    throw malformed_input();
+  }
+  if (!mask || !value) {
+    throw malformed_input();
+  }
+  Dtype index_dtype = dtypeOfIndices(indices);
+  if (index_dtype.lanes() != mask->dtype().lanes()) {
+    throw malformed_input();
+  }
+  if (index_dtype.lanes() != value->dtype().lanes()) {
+    throw malformed_input();
+  }
+  */
+}
+
+Store* Store::make(
+    const Buffer& buffer,
+    const std::vector<ExprHandle>& indices,
+    const ExprHandle& value,
+    const ExprHandle& mask) {
+  return new Store(
+      buffer, ExprHandleVectorToExprVector(indices), value.node(), mask.node());
+}
+
+Store* Store::make(
+    const BufHandle& buf,
+    const std::vector<ExprHandle>& indices,
+    const ExprHandle& value,
+    const ExprHandle& mask) {
+  return new Store(
+      buf.node(),
+      ExprHandleVectorToExprVector(indices),
+      value.node(),
+      mask.node());
+}
+
+Store* Store::make(
+    const BufHandle& buf,
+    const std::vector<ExprHandle>& indices,
+    const ExprHandle& value) {
+  return new Store(
+      buf.node(),
+      ExprHandleVectorToExprVector(indices),
+      value.node(),
+      ExprHandle(1).node());
+}
+
+const Expr* flattened_index(
+    const std::vector<const Expr*>& dims,
+    const std::vector<const Expr*>& indices) {
+  // Handle already flattened indices first
+  if (indices.size() == 1) {
+    return indices[0];
+  }
+
+  size_t ndim = dims.size();
+  if (ndim != indices.size()) {
+    throw malformed_input();
+  }
+  std::vector<ExprHandle> strides(ndim);
+  for (size_t i = 0; i < ndim; i++) {
+    if (i == ndim - 1) {
+      strides[i] = ExprHandle(1);
+      continue;
+    }
+    ExprHandle stride = ExprHandle(dims[i + 1]);
+    for (size_t j = i + 2; j < ndim; j++) {
+      stride = stride * ExprHandle(dims[j]);
+    }
+    strides[i] = stride;
+  }
+
+  ExprHandle total_index = int32_t{0};
```

Comment:
Why "int32_t{0}"?

Reply:
The code for computing `total_index` was there before, I copied it almost verbatim :)